Fix #653 : Landscape add profile

MohamedMedhat1998 opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
 -Fixes #653 : Landscape add profile
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
## Before
The **AddProfileActivity** was locked to portrait mode, when you rotate the device, nothing happens

![Screenshot_2020-03-08-09-08-18](https://user-images.githubusercontent.com/8533363/76162441-837f8980-6146-11ea-9441-f16378f1f5bf.jpg)

## After
The **AddProfileActivity** now supports the landscape mode, and that is how it looks after the modification

![Screenshot_2020-03-08-14-07-22](https://user-images.githubusercontent.com/8533363/76162554-7fa03700-6147-11ea-86cf-df4232f685ee.png)

![Screenshot_2020-03-08-14-08-27](https://user-images.githubusercontent.com/8533363/76162559-8464eb00-6147-11ea-9b3d-3e7b2a69638e.png)

![Screenshot_2020-03-08-14-08-34](https://user-images.githubusercontent.com/8533363/76162561-87f87200-6147-11ea-92f9-ac7c2b95f6cb.png)

![Screenshot_2020-03-08-14-08-37](https://user-images.githubusercontent.com/8533363/76162564-8af36280-6147-11ea-8547-63b99d830149.png)


## Checklist
- [x] Lowfi Landscape
- [x] Test cases for Landscape

<!-- Please tick the relevant boxes by putting an "x" in them. -->

MohamedMedhat1998 committed:
-Added a layout for the landscape mode for "AddProfileActivity"
-Added a custom test class for the constraint layout
-Used this class to build custom test cases

rt4914 reviewed (changes requested):
1. In manifest the `AddProfileActivity` is still fixed to portrait mode.
2. In test cases you don't have to write test cases around UI/Constraints instead we will be writing test cases around the functionality. For now , do one thing remove those test cases and the new file that you have introduced. And make sure that the orientation change in `AddProfile` is working correctly once that is done, we can write test cases.

MohamedMedhat1998 committed:
-Modified "android:screenOrientation" property in the manifest for the "AddProfileActivity" so that it is not tied to the portrait mode
-Removed the "ConstraintMatcher" class -A custom class for testing-
-Removed the test cases that tests the constraint in the "AddProfileActivityTest"

MohamedMedhat1998 commented:
@rt4914 I have done the required changes, please check.

veena14cs commented on app/src/main/res/layout-land/add_profile_activity.xml:
```diff
@@ -0,0 +1,194 @@
+          android:id="@+id/input_confirm_pin"
+          android:layout_width="match_parent"
+          android:layout_height="wrap_content"
+          android:layout_marginStart="@dimen/margin_32"
```
Use direct values instead of dimen values for margins.

MohamedMedhat1998 replied:
I didn't write this by myself. I copied it from the portrait mode layout file, so I think I don't need to change them, am I correct?

veena14cs commented on app/src/main/res/layout-land/add_profile_activity.xml:
```diff
@@ -0,0 +1,194 @@
+          android:id="@+id/input_pin"
+          android:layout_width="match_parent"
+          android:layout_height="wrap_content"
+          android:layout_marginStart="@dimen/margin_32"
```
Ditto.

veena14cs commented on app/src/sharedTest/java/org/oppia/app/profile/AddProfileActivityTest.kt:
```diff
@@ -329,6 +329,8 @@ class AddProfileActivityTest {
     }
   }
 
+
```
Remove blank spaces.

MohamedMedhat1998 replied:
Ok, I'll do it

nikitamarysolomanpvt commented:
@MohamedMedhat1998  Please add test-cases

MohamedMedhat1998 committed:
-Removed the white spaces from the "AddProfileActivityTest" class

rt4914 reviewed (changes requested):
Please add test cases for this implementation. Your test cases will be something like this:

Whatever test cases are there currently in `AddProfileActivityTest`, similar test cases for landscape.

Along with that one special test cases will as follows:
1. Type user name in portrait mode and do orientation change, check if after orientation change the typed text is still there or not. This will be done for all there exit-text separately and one test case in which text will be written in all three edittext, do the orientation change and then check that all three edittext contains text.
Also, you click on check box in portrait mode, do orientation change, the checkbox should still be selected and the bottom 2 edit texts should be visible.

MohamedMedhat1998 commented:
@rt4914 Should I fix something in "AddProfileActivity" in order to pass a test case that I wrote recently, or someone else should fix that?

seanlip commented:
If the current behaviour is wrong, you should fix that as part of writing your test.

If the current behaviour is correct, then perhaps your test needs to be revisited.

Does that help?

MohamedMedhat1998 committed:
-Completed all the test cases for the landscape mode for "AddProfileActivity"

MohamedMedhat1998 commented:
@rt4914 Please check the PR, I added test cases for the landscape mode.  
Please check them before I try to modify the `AddProfileActivity` to pass the tests that fail

rt4914 reviewed (changes requested):
@MohamedMedhat1998 Actually the approach here is incorrect. First we need to make sure that the code is working correctly in terms of orientation change and if not then you will need to modify the code first. Once the app is working correctly then we should write test cases.


We need to make sure that following conditions are met in app. Once this is done then only we can work on test cases.

Type user name in portrait mode and do orientation change, check if after orientation change the typed text is still there or not (If not, then use onSavedInstance to fix this). This will be done for all there exit-text separately and one test case in which text will be written in all three edittext, do the orientation change and then check that all three edittext contains text.
Also, you click on check box in portrait mode, do orientation change, the checkbox should still be selected and the bottom 2 edit texts should be visible.

MohamedMedhat1998 committed:
-Manually saved the state of the "AddProfileActivity" (now the landscape mode works as expected)

rt4914 commented on app/src/main/java/org/oppia/app/profile/AddProfileViewModel.kt:
```diff
@@ -13,6 +13,11 @@ class AddProfileViewModel @Inject constructor() : ObservableViewModel() {
   val confirmPinErrorMsg = ObservableField("")
   val nameErrorMsg = ObservableField("")
   val createPin = ObservableField(false)
+  var savedName = ""
```
Instead of `var` can make all these filelds as ObservableField similar to `createPin`, `nameErrorMsg` etc.

For example:
`val savedName = ObservableField<String>("")`

MohamedMedhat1998 replied:
I'll try that

rt4914 commented on app/src/sharedTest/java/org/oppia/app/profile/AddProfileActivityTest.kt:
```diff
@@ -91,6 +94,30 @@ class AddProfileActivityTest {
+  @Test
+  fun testAddProfileActivity_configurationChange_inputName_clickCreate_checkOpensProfileActivity() {
+    ActivityScenario.launch(AddProfileActivity::class.java).use {
+      onView(ViewMatchers.isRoot()).perform(OrientationChangeAction.orientationLandscape())
```
Make is simple by importing the Classes.Functions direclty.
`onView(isRoot()).perform(orientationLandscape())`

This needs to be done in all test cases and also make sure that unused imports are removed

MohamedMedhat1998 replied:
Got what you mean, I'll fix that.

rt4914 commented on app/src/main/java/org/oppia/app/profile/AddProfileActivity.kt:
```diff
@@ -40,4 +40,8 @@ class AddProfileActivity : InjectableAppCompatActivity() {
     addProfileFragmentPresenter.handleOnActivityResult(requestCode, resultCode, data)
   }
+
+  override fun onRestoreInstanceState(savedInstanceState: Bundle?) {
```
Remove this function as it is not doing anything.

MohamedMedhat1998 replied:
Actually, it stops the system from saving the state and restoring it. This allows me to handle the state manually on the configuration change.  
I had to do it this way because there is a bug related to the view binding. If I type a name only in the `input_name` and make a configuration change, the first 3 letters are input in both `input_pin` and `input_confirm_pin`. I think that happens because of the custom view which has the same id for the `EditText` in each view.  
I'm not sure how `ViewBinding` handles this, but I think it is related to it.

rt4914 commented on app/src/main/java/org/oppia/app/profile/AddProfileViewModel.kt:
```diff
@@ -13,6 +13,11 @@ class AddProfileViewModel @Inject constructor() : ObservableViewModel() {
   val nameErrorMsg = ObservableField("")
   val createPin = ObservableField(false)
+  var savedName = ""
+  var savedPin = ""
```
All these values are actually not getting used in xml file. Suggest using the concept of data bining to set values via viewmodel. Similar to  `nameErrorMsg`

MohamedMedhat1998 replied:
Ok I'll check that

rt4914 commented on app/src/main/java/org/oppia/app/profile/AddProfileActivityPresenter.kt:
```diff
@@ -79,6 +80,17 @@ class AddProfileActivityPresenter @Inject constructor(
     addTextChangedListeners(binding)
     addButtonListeners(binding)
+
+    binding.inputName.setInput(profileViewModel.savedName)
```
Instead of setting data to xml using binding, use viewModel. Check how `createPin` is getting used in xml from `ViewModel`. In this presenter you can just set data in viewmodel and then viewmodel will set data to the xml file.

MohamedMedhat1998 replied:
ok

rt4914 reviewed (commented):
The implementation is working but the approach is incorrect. Please fix all the changes suggested in this review. Also, we do need some more test cases but I would suggest lets just focus on entire implementation first, we can add new test cases later on.

Thanks.

rt4914 commented:
@MohamedMedhat1998 Haven't seen any update on this PR. Are you stuck or busy?

MohamedMedhat1998 commented:
@rt4914 I had some troubles but everything is good now.
I'm working on the changes right now.

MohamedMedhat1998 commented:
@rt4914 I'm stuck at some point.  
When I type for example "Mohamed" in the `input_name` then I rotate the device  
The `input_pin` and `input_confirm_pin` are filled with "Moh". I can't figure out why this happens.

MohamedMedhat1998 commented:
@rt4914  
I could figure out finally why this happens, but it is something beyond my abilities.  
It happens because the `EditText` tries to save its state during configuration changes. It depends on the id to achieve that and we are using custom view here `ProfileInputView` which uses an `EditText` with the same `id` in each instance, so it updates all the instances with the same text on configuration changes.  
I have created a [new branch in this project](https://github.com/MohamedMedhat1998/Playground/tree/custom-view-same-id-save-state) and created a custom view and tested it and I faced the same problem.  
So, it would be difficult -_At least for me_- to save the state on configuration change while having multiple instances of the custom view using the `DataBinding`.  
Here is a gif of what is happening  

![save_state](https://user-images.githubusercontent.com/8533363/76980702-31a6e280-6942-11ea-8cb4-50015fb5d883.gif)

So, can you merge this PR please or help me solve this problem?  
If it is not possible, can you assign me an easier issue please?  
I need to have at least 2 merged PR to be eligible to apply to gsoc and that PR took a lot of time :(

rt4914 commented:
> @rt4914
> I could figure out finally why this happens, but it is something beyond my abilities.
> It happens because the `EditText` tries to save its state during configuration changes. It depends on the id to achieve that and we are using custom view here `ProfileInputView` which uses an `EditText` with the same `id` in each instance, so it updates all the instances with the same text on configuration changes.
> I have created a [new branch in this project](https://github.com/MohamedMedhat1998/Playground/tree/custom-view-same-id-save-state) and created a custom view and tested it and I faced the same problem.
> So, it would be difficult -_At least for me_- to save the state on configuration change while having multiple instances of the custom view using the `DataBinding`.
> Here is a gif of what is happening
> 
> ![save_state](https://user-images.githubusercontent.com/8533363/76980702-31a6e280-6942-11ea-8cb4-50015fb5d883.gif)
> 
> So, can you merge this PR please or help me solve this problem?
> If it is not possible, can you assign me an easier issue please?
> I need to have at least 2 merged PR to be eligible to apply to gsoc and that PR took a lot of time :(

@MohamedMedhat1998 It is completely fine if it was hard to finish, as actually it is a hard task. Anyways, lets close this PR and assign you some task which can help you grow step-by-step.

I have assigned you https://github.com/oppia/oppia-android/issues/772
